File: TextShortener.py
class TextShortener:
    def split_text_into_words(self, input_text):
        words = []
        current_word = []

        for char in input_text:
            if char.isalnum() or char in ("'", '-'):
                current_word.append(char)
            else:
                if current_word:
                    words.append(''.join(current_word))
                    current_word.clear()

        if current_word:
            words.append(''.join(current_word))

        return words

    def split_word_into_syllables(self, word):
        lower_word = word.lower()
        vowel_positions = []

        vowels = set('аеёиоуыэюяaeiouy')

        for i in range(len(lower_word)):
            if lower_word[i] in vowels:
                vowel_positions.append(i)

        if not vowel_positions:
            return [word]

        result = []
        prev_pos = 0

        for i in range(len(vowel_positions)):
            current_vowel_pos = vowel_positions[i]
            if i > 0:
                split_pos = (current_vowel_pos + vowel_positions[i - 1] + 1) // 2
                if word[split_pos] in ('ь', 'ъ'):
                    split_pos += 1
                result.append(word[prev_pos:split_pos])
                prev_pos = split_pos

        result.append(word[prev_pos:])
        return result

    def shorten_one_word(self, input_word, max_length, symbol="-"):
        if not input_word or max_length <= 0:
            return ''
        syllables = self.split_word_into_syllables(input_word)
        if len(syllables) < 3:
            return f"{input_word[:max_length]}~0"

        max_length -= len(symbol)

        full_string = ''.join(syllables)
        l = -1
        while True:
            if len(syllables) < 3:
                return f"{input_word[:max_length-1]}~0"

            l = len(syllables) // 2
            syllables.pop(l)
            full_string = ''.join(syllables)
            if len(full_string) <= max_length:
                break

        syllables.insert(l, symbol)
        full_string = ''.join(syllables)
        return full_string

    def shorten_two_words(self, input_word, max_length, afterwords=1):
        max_length -= (1 + len(str(abs(afterwords))))
        if len(input_word) < -max_length:
            return f"{input_word}~{afterwords}"
        shortened = self.shorten_one_word(input_word, max_length)
        return f"{shortened}~{afterwords}"

    def trim_string(self, input_text, max_length, space=" ", symbol=".."):
        if not input_text or max_length <= 0:
            return ''
        words = self.split_text_into_words(input_text)
        if not words:
            return ''

        if len(words) == 1:
            return self.shorten_one_word(words[0], max_length)
        if len(words) == 2:
            return self.shorten_two_words(words[0], max_length)

        full_string = space.join(words)
        if len(full_string) <= max_length:
            return full_string

        base_count = len(words)
        max_words_length = max_length - len(space) * 2 + len(symbol)

        l = -1
        while True:
            if len(words) == 2:
                return self.shorten_two_words(words[0], max_length, base_count - 1)

            l = len(words) // 2
            words.pop(l)
            full_string = space.join(words)
            if len(full_string) <= max_words_length:
                break

        words.insert(l, symbol)
        full_string = space.join(words)
        return full_string

    def test(self):
        print( self.trim_string("Ехал грека через реку, видит грека в реке рак", 30) )
        print( self.trim_string("Череззаборногузадерищенский кактусообразный параллепипед", 45) )
        print( self.trim_string("Череззаборногузадерищенский кактусообразный параллепипед", 22) )
        print( self.trim_string("Череззаборногузадерищенский параллепипед", 22) )
        print( self.trim_string("Преподавательница", 12) )
        
if __name__ == "__main__":
    t = TextShortener();
    t.test()
    pass

File: TextShortener.cs
using System;
using System.Collections.Generic;
using System.Text;

public static class TextShorterner
{
    /// <summary>
    ///     Делим предложение на слова
    /// </summary>
    /// <param name="input"></param>
    /// <returns></returns>
    internal static List<string> SplitTextIntoWords(string input)
    {
        List<string> words = new List<string>();
        StringBuilder currentWord = new StringBuilder();

        foreach (char c in input)
        {
            if (char.IsLetterOrDigit(c) || c == '\'' || c == '-')
            {
                currentWord.Append(c);
            }
            else
            {
                if (currentWord.Length > 0)
                {
                    words.Add(currentWord.ToString());
                    currentWord.Clear();
                };
            };
        };

        if (currentWord.Length > 0)
            words.Add(currentWord.ToString());

        return words;
    }

    /// <summary>
    ///     Делим слово по слогам
    /// </summary>
    /// <param name="word"></param>
    /// <returns></returns>
    internal static List<string> SplitWordIntoSyllables(string word)
    {
        string lowerWord = word.ToLower();
        List<int> vowelPositions = new List<int>();

        HashSet<char> vowels = new HashSet<char> { 
            'а', 'е', 'ё', 'и', 'о', 'у', 'ы', 'э', 'ю', 'я', // RUS
            'a', 'e', 'i', 'o', 'u' , 'y' }; // ENG

        for (int i = 0; i < lowerWord.Length; i++)
            if (vowels.Contains(lowerWord[i])) 
                vowelPositions.Add(i);

        if (vowelPositions.Count == 0)
            return new List<string> { word };

        List<string> result = new List<string>();
        int prevPos = 0;

        for (int i = 0; i < vowelPositions.Count; i++)
        {
            int currentVowelPos = vowelPositions[i];
            if (i > 0)
            {
                int splitPos = (currentVowelPos + vowelPositions[i - 1] + 1) / 2;
                if (word[splitPos] == 'ь' || word[splitPos] == 'ъ') splitPos++;
                result.Add(word.Substring(prevPos, splitPos - prevPos));
                prevPos = splitPos;
            };
        };
        result.Add(word.Substring(prevPos));

        return result;
    }

    /// <summary>
    ///     Сокращаем одно слово через дефис
    /// </summary>
    /// <param name="input"></param>
    /// <param name="maxLength"></param>
    /// <returns></returns>
    internal static string ShortenOneWord(string input, int maxLength, string symbol = "-")
    {
        if (string.IsNullOrEmpty(input) || maxLength <= 0) return string.Empty;
        List<string> syllables = SplitWordIntoSyllables(input);
        if (syllables.Count < 3) return $"{input.Substring(0, maxLength)}~0";

        maxLength -= symbol.Length;

        string fullString = string.Join("", syllables);
        int l = -1;
        do
        {
            if (syllables.Count < 3)
                return $"{input.Substring(0,--maxLength)}~0";

            l = syllables.Count / 2;
            syllables.RemoveAt(l);
            fullString = string.Join("", syllables);
        } 
        while (fullString.Length > maxLength);

        syllables.Insert(l, symbol);
        fullString = string.Join("", syllables);
        return fullString;
    }

    /// <summary>
    ///     Сокращаем два слова
    /// </summary>
    /// <param name="input"></param>
    /// <param name="maxLength"></param>
    /// <param name="afterwords"></param>
    /// <returns></returns>
    internal static string ShortenTwoWords(string input, int maxLength, int afterwords = 1)
    {
        maxLength -= (1 + Math.Abs(afterwords).ToString().Length);
        if (input.Length < -maxLength) return $"{input}~{afterwords}";
        string shorten = ShortenOneWord(input, maxLength);
        return $"{shorten}~{afterwords}";
    }

    public static string TrimString(string input, int maxLength, string space = " ", string symbol = "..")
    {
        if (string.IsNullOrEmpty(input) || maxLength <= 0) return string.Empty;
        List<string> words = SplitTextIntoWords(input);
        if (words.Count == 0) return string.Empty;

        if (words.Count == 1) return ShortenOneWord(words[0], maxLength);
        if (words.Count == 2) return ShortenTwoWords(words[0], maxLength);

        string fullString = string.Join(space, words);
        if (fullString.Length <= maxLength) return fullString;

        int baseCount = words.Count;
        int maxWordsLength = maxLength - space.Length * 2 + symbol.Length;

        int l = -1;
        do
        {
            if(words.Count == 2)
                return ShortenTwoWords(words[0], maxLength, baseCount-1);

            l = words.Count / 2;
            words.RemoveAt(l);            
            fullString = string.Join(space, words);
        } 
        while (fullString.Length > maxWordsLength);

        words.Insert(l, symbol);
        fullString = string.Join(space, words);
        return fullString;
    }


    internal static void Test()
    {
        string bs1 = TextShorterner.TrimString("Ехал грека через реку, видит грека в реке рак", 30);
        string bs2 = TextShorterner.TrimString("Череззаборногузадерищенский кактусообразный параллепипед", 45);
        string bs3 = TextShorterner.TrimString("Череззаборногузадерищенский кактусообразный параллепипед", 22);
        string bs4 = TextShorterner.TrimString("Череззаборногузадерищенский параллепипед", 22);
        string bs5 = TextShorterner.TrimString("Преподавательница", 12);
    }
}